enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,610 @@
+***GCP Labels*** 
+
+Add a new field `resourceLabels` to `.status.platformStatus.gcp` of the
+`infrastructure.config.openshift.io` type. Labels included in the `resourceLabels` field
+will be applied to new resources created for the cluster by the in-cluster operators.
+
+The `resourceLabels` field in `status.platformStatus.gcp` will be populated by the
+installer using the entries from `platform.gcp.userLabels` field of `install-config`.
+
+`status.platformStatus.gcp.resourceLabels` field of `infrastructure.config.openshift.io`
+is immutable.
+
+OpenShift operators(Cluster Infrastructure, Storage) that create GCP resources will apply
+these labels to all GCP resources they create.
+
+Below list of terraform GCP APIs to create resources should be updated to add user
+defined labels and as well the OpenShift default label in the installer component.
+`google_storage_bucket, google_compute_instance, google_dns_managed_zone, google_compute_image, google_compute_forwarding_rule`
+
+And below list of terraform GCP APIs support labeling resources in beta version.
+`google_compute_address`
+
+Below list of GCP Resources does not support labeling.
+`VPC Network, VM Instance Group, Subnetwork, Target Pool, DNS Zone, Records, Router, NAT Rules, Backend Service, Health Check`
+
+API update example:
+A local variable should be defined, which merges the default label and the user
+defined GCP labels, which should be referred in the GCP resource APIs.
+``` terraform
+locals {
+  labels = merge(
+    {
+      "kubernetes-io-cluster-${var.cluster_id}" = "owned"
+    },
+    var.gcp_extra_labels,
+  )
+}
+
+resource "google_storage_bucket" "ignition" {
+  ...
+  labels = local.labels
+}
+```
+
+For resources supporting labels in beta version below changes are required.
+``` terraform
+provider "googlebeta" {
+  credentials = var.gcp_service_account
+  project     = var.gcp_project_id
+  region      = var.gcp_region
+}
+
+resource "google_compute_address" "cluster_ip" {
+  provider = googlebeta
+  ...
+  labels = local.labels
+}
+```
+
+The label of the form `kubernetes-io-cluster-<cluster_id>:owned` added by OpenShift, where
+cluster_id is a string generated by concatenating user inputted cluster name and a random
+string will be limited to a maximum length of 27 characters by trimming long cluster name
+to 21 characters.
+
+***GCP Tags*** 
+
+Add a new field `resourceTags` to `.status.platformStatus.gcp` of the
+`infrastructure.config.openshift.io` type. Tags included in the `resourceLabels` field
+will be applied to new resources created for the cluster by the in-cluster operators.
+
+The `resourceTags` field in `status.platformStatus.gcp` will be populated by the
+installer using the entries from `platform.gcp.userTags` field of `install-config`.
+
+`status.platformStatus.gcp.resourceTags` field of `infrastructure.config.openshift.io`
+is immutable.
+
+OpenShift operators(Cluster Infrastructure, Storage) that create GCP resources will apply
```

Comment:
Does Cluster API Provider GCP support applying tags?

Reply:
I think, it's not supported. I went through the reconciler code and couldn't find any references for user-defined tags, but they support defining network tags, which is present in machine-api as well.

Reply:
It will need to be added upstream before we can accept it downstream in that case

Reply:
Changes in GCPMachineProviderSpec is required for users to configure additional tags apart from those provided during cluster creation in install-config. Since VMs will tagged with tags provided by user during installation and extending GCPMachineProviderSpec to support tags require upstream and due to the time limit, would it be acceptable to add this feature in GA and in TP it will be a limitation?

Reply:
Added as limitation for TP.